_posts/2019-06-25-dynamic_linear_models_tfprobability/dynamic_linear_models_tfprobability.Rmd

```diff
@@ -0,0 +1,805 @@
+
+Assuming this relationship does not change over time, we can easily use `lm` to illustrate this. Following Petris et al. in zooming in on IBM as the asset under study, we have
+
+```{r}
+# excess returns of the asset under study
+ibm <- df$IBM - df$RKFREE
+# market excess returns
+x <- df$MARKET - df$RKFREE
+
+fit <- lm(ibm ~ x)
+summary(fit)
+```
+
+
+```
+Call:
+lm(formula = ibm ~ x)
+
+Residuals:
+     Min       1Q   Median       3Q      Max 
+-0.11850 -0.03327 -0.00263  0.03332  0.15042 
+
+Coefficients:
+              Estimate Std. Error t value Pr(>|t|)    
+(Intercept) -0.0004896  0.0046400  -0.106    0.916    
+x            0.4568208  0.0675477   6.763 5.49e-10 ***
+---
+Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
+
+Residual standard error: 0.05055 on 118 degrees of freedom
+Multiple R-squared:  0.2793,	Adjusted R-squared:  0.2732 
+F-statistic: 45.74 on 1 and 118 DF,  p-value: 5.489e-10
+```
+
+
+So IBM is found to be a conservative investment, the slope being ~ 0.5. But is this relationship stable over time?
+
+Let's turn to `tfprobability` to investigate.
+
+We want to use this example to demonstrate two essential applications of DLMs: obtaining smoothing and/or filtering estimates of the coefficients, as well as forecasting future values. So unlike Petris et al., we divide the dataset into a training and a testing part:^[Forecasting functionality is not provided by `dlm` for dynamic linear models, see `?dlmForecast`].
+
+
+```{r}
+# zoom in on ibm
+ts <- ibm %>% matrix()
+# forecast 12 months
+n_forecast_steps <- 12
+ts_train <- ts[1:(length(ts) - n_forecast_steps), 1, drop = FALSE]
+
+# make sure we work with float32 here
+ts_train <- tf$cast(ts_train, tf$float32)
+ts <- tf$cast(ts, tf$float32)
+```
+
+
+We now construct the model. [sts_dynamic_linear_regression()](https://rstudio.github.io/tfprobability/reference/sts_dynamic_linear_regression.html) does what we want: 
+
+```{r}
+# define the model on the complete series
+linreg <- ts %>%
+  sts_dynamic_linear_regression(
+    design_matrix = cbind(rep(1, length(x)), x) %>% tf$cast(tf$float32)
+  )
+```
+
+
+We pass it the column of excess market returns, plus a column of ones, following Petris et al.. Alternatively, we could center the single predictor -- this would work just as well. 
+
+How are we going to train this model? Method-wise, we have a choice between variational inference (VI) and Hamiltonian Monte Carlo (HMC). We will see both. The second question is: Are we going to use graph mode or eager mode? As of today, for both VI and HMC, it is safest -- and fastest -- to run in graph mode, so this is the only technique we show. In a few weeks, or months, we should be able to prune a lot of `sess$run()`s from the code!
```

Comment:
> In a few weeks, or months, we should be able to prune a lot of `sess$run()`s from the code!

Is this referencing upcoming changes in one of the packages? You've been using "we" to describe what you're doing in this post (with the reader accompanying you), so the "we" here is a bit confusing.